In [1]:
import pandas as pd

In [2]:
dt = pd.read_csv("../Dataset/clean_medicine_data.csv")

In [3]:
dt.head(5)

,id,name,price,Is_discontinued,manufacturer_name,type,pack_size_label,medicine_desc,side_effects,drug_interactions,composition,cleaned_name,document
0,1,Augmentin 625 Duo Tablet,223,False,glaxo smithkline pharmaceuticals ltd,allopathy,strip of 10 tablets,Augmentin 625 Duo Tablet is a penicillin-type ...,"Vomiting,Nausea,Diarrhea","{""drug"": [], ""brand"": [], ""effect"": []}",amoxycillin (500mg) clavulanic acid (125mg),augmentin 625 duo tablet,"\n Medicine : augmentin 625 duo tablet,\n ..."
1,2,Azithral 500 Tablet,132,False,alembic pharmaceuticals ltd,allopathy,strip of 5 tablets,Azithral 500 Tablet is an antibiotic used to t...,"Vomiting,Nausea,Abdominal pain,Diarrhea","{""drug"": [""Astemizole"", ""Ebastine"", ""Mizolasti...",azithromycin (500mg),azithral 500 tablet,"\n Medicine : azithral 500 tablet,\n Com..."
2,3,Ascoril LS Syrup,118,False,glenmark pharmaceuticals ltd,allopathy,bottle of 100 ml syrup,Ascoril LS Syrup is a combination medicine use...,"Nausea,Vomiting,Diarrhea,Upset stomach,Stomach...","{""drug"": [], ""brand"": [], ""effect"": []}",ambroxol (30mg/5ml) levosalbutamol (1mg/5ml),ascoril ls syrup,"\n Medicine : ascoril ls syrup,\n Compos..."
3,4,Allegra 120mg Tablet,218,False,sanofi india ltd,allopathy,strip of 10 tablets,Allegra 120mg Tablet is an anti-allergy medici...,"Headache,Drowsiness,Dizziness,Nausea","{""drug"": [""Aprepitant"", ""Fosaprepitant"", ""Maga...",fexofenadine (120mg),allegra 120mg tablet,"\n Medicine : allegra 120mg tablet,\n Co..."
4,5,Avil 25 Tablet,10,False,sanofi india ltd,allopathy,strip of 15 tablets,Avil 25 Tablet is an antiallergic medication u...,Sleepiness,"{""drug"": [""Zafirlukast"", ""Alprazolam"", ""Betahi...",pheniramine (25mg),avil 25 tablet,"\n Medicine : avil 25 tablet,\n Composit..."


#### We will create Dictionaries to map name to index and vice-versa for fast lookup

In [4]:
name_to_index = {
    name:idx for idx,name in enumerate(dt['cleaned_name'])
}

index_to_name = {
    idx:name for idx,name in enumerate(dt['cleaned_name'])
}

In [2]:
def get_composition(medicine:str):
    medicine = medicine.lower().strip()

    if medicine not in name_to_index:
        return None
    
    idx = name_to_index[medicine]
    return dt.loc[idx, 'composition']

In [3]:
def find_exact_substitutes(medicine:str, top_k=5):
    comp = get_composition(medicine)

    if comp is None:
        return "Medicine not found."

    substitutes = dt[dt['composition'] == comp]
    substitutes = substitutes.sort_values("price")

    return substitutes[[
        "name",
        "manufacturer_name",
        "price",
        "pack_size_label",
        "composition"
    ]].head(top_k)

In [7]:
find_exact_substitutes("Allegra 120mg Tablet")

,name,manufacturer_name,price,pack_size_label,composition
87339,Faxodas 120mg Tablet,shreya life sciences pvt ltd,19,strip of 6 tablets,fexofenadine (120mg)
236774,Zelgra 120 Tablet,zeelab pharmacy pvt ltd,22,strip of 10 tablets,fexofenadine (120mg)
1265,Alert 120mg Tablet,alkem laboratories ltd,28,strip of 10 tablets,fexofenadine (120mg)
60041,DavaIndia Fexofenadine 120mg Tablet,davaindia generic pharmacy,29,strip of 10 tablets,fexofenadine (120mg)
4458,Allegix 120mg Tablet,intas pharmaceuticals ltd,30,strip of 6 tablets,fexofenadine (120mg)


In [8]:
find_exact_substitutes("benadryl syrup")

,name,manufacturer_name,price,pack_size_label,composition
103810,Intadryl Syrup,intas pharmaceuticals ltd,38,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
156341,Odaril Syrup,moraceae pharmaceuticals pvt ltd,43,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
207575,Tussberry-N Syrup,acinta pharmaceuticals pvt ltd,51,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
145633,New Brethese Syrup,galpha laboratories ltd,68,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
193007,Suntuss CF Syrup,suncare formulations pvt ltd,70,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...


### Similarity Matching (TF-IDF)

In [4]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

tfidf = TfidfVectorizer()
comp_vectors = tfidf.fit_transform(dt['composition'])

NameError: name 'dt' is not defined

In [ ]:
def find_similar_substitutes(medicine:str, top_k=5):
    idx = name_to_index.get(medicine.lower())

    if idx is None:
        return "Medicine not found."
    
    query_vec = comp_vectors[idx]

    similarities = cosine_similarity(query_vec, comp_vectors)[0]

    dt["similarity"] = similarities

    result = dt.sort_values(['similarity', 'price'],ascending=[False,True])

    return result[[
        "name",
        "manufacturer_name",
        "price",
        "pack_size_label",
        "composition",
        "similarity"
    ]].head(20)

In [11]:
find_similar_substitutes("benadryl syrup")

,name,manufacturer_name,price,pack_size_label,composition,similarity
103810,Intadryl Syrup,intas pharmaceuticals ltd,38,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
156341,Odaril Syrup,moraceae pharmaceuticals pvt ltd,43,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
207575,Tussberry-N Syrup,acinta pharmaceuticals pvt ltd,51,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
145633,New Brethese Syrup,galpha laboratories ltd,68,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
193007,Suntuss CF Syrup,suncare formulations pvt ltd,70,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
197356,Super Tus Syrup,worth medicines pvt ltd,80,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
33118,COF-RYL Cough Syrup,cipla ltd,82,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
36420,Cofryl Syrup,cipla ltd,82,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
32725,Cofsils Cough Syrup,cipla health ltd,95,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
199137,Suncaryl Syrup,suncare formulations pvt ltd,110,bottle of 150 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000


### Embeddings

In [5]:
from sentence_transformers import SentenceTransformer

c:\ML Projects\AIPharma\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [13]:
dt.head(5)

,id,name,price,Is_discontinued,manufacturer_name,type,pack_size_label,medicine_desc,side_effects,drug_interactions,composition,cleaned_name,document,similarity
0,1,Augmentin 625 Duo Tablet,223,False,glaxo smithkline pharmaceuticals ltd,allopathy,strip of 10 tablets,Augmentin 625 Duo Tablet is a penicillin-type ...,"Vomiting,Nausea,Diarrhea","{""drug"": [], ""brand"": [], ""effect"": []}",amoxycillin (500mg) clavulanic acid (125mg),augmentin 625 duo tablet,"\n Medicine : augmentin 625 duo tablet,\n ...",0.000000
1,2,Azithral 500 Tablet,132,False,alembic pharmaceuticals ltd,allopathy,strip of 5 tablets,Azithral 500 Tablet is an antibiotic used to t...,"Vomiting,Nausea,Abdominal pain,Diarrhea","{""drug"": [""Astemizole"", ""Ebastine"", ""Mizolasti...",azithromycin (500mg),azithral 500 tablet,"\n Medicine : azithral 500 tablet,\n Com...",0.000000
2,3,Ascoril LS Syrup,118,False,glenmark pharmaceuticals ltd,allopathy,bottle of 100 ml syrup,Ascoril LS Syrup is a combination medicine use...,"Nausea,Vomiting,Diarrhea,Upset stomach,Stomach...","{""drug"": [], ""brand"": [], ""effect"": []}",ambroxol (30mg/5ml) levosalbutamol (1mg/5ml),ascoril ls syrup,"\n Medicine : ascoril ls syrup,\n Compos...",0.229844
3,4,Allegra 120mg Tablet,218,False,sanofi india ltd,allopathy,strip of 10 tablets,Allegra 120mg Tablet is an anti-allergy medici...,"Headache,Drowsiness,Dizziness,Nausea","{""drug"": [""Aprepitant"", ""Fosaprepitant"", ""Maga...",fexofenadine (120mg),allegra 120mg tablet,"\n Medicine : allegra 120mg tablet,\n Co...",0.000000
4,5,Avil 25 Tablet,10,False,sanofi india ltd,allopathy,strip of 15 tablets,Avil 25 Tablet is an antiallergic medication u...,Sleepiness,"{""drug"": [""Zafirlukast"", ""Alprazolam"", ""Betahi...",pheniramine (25mg),avil 25 tablet,"\n Medicine : avil 25 tablet,\n Composit...",0.000000


In [6]:
model = SentenceTransformer("all-MiniLM-L6-v2")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 1080.40it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [16]:
embeddings = model.encode(
    dt['document'].tolist(),
    show_progress_bar=True,
    batch_size=64
)

Batches: 100%|██████████| 3807/3807 [25:16<00:00,  2.51it/s]


In [ ]:
import numpy as np
import faiss

new_embeddings = np.array(embeddings).astype("float32")
faiss.normalize_L2(new_embeddings)  # Normalize the vectors

#### Create Flat Index (Inner Product)

In [30]:
dimension = new_embeddings.shape[1]

index = faiss.IndexFlatIP(dimension)
index.add(new_embeddings)

#### Save Index 

In [ ]:
faiss.write_index(index, "../data/medicine_cosine_flat.index") 
dt.to_pickle("../data/medicine_metadata.pkl")

#### To load it later run this
>index = faiss.read_index("../Data/medicine_cosine_flat.index")

>df = pd.read_pickle("../Data/medicine_metadata.pkl")

In [35]:
def semantic_search(query, top_k=5):

    query_embedding = model.encode([query]).astype("float32")

    # normalize query
    faiss.normalize_L2(query_embedding)

    scores, indices = index.search(query_embedding, top_k)

    results = dt.iloc[indices[0]].copy()
    results["similarity"] = scores[0]

    return results[
        ["name", "composition", "manufacturer_name", "price", "similarity"]
    ]

In [38]:
semantic_search("benadryl syrup")

,name,composition,manufacturer_name,price,similarity
24840,Benadryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,johnson & johnson ltd,129,0.748119
25019,Benadryl DR Kids Syrup,dextromethorphan hydrobromide (7.5mg/5ml),johnson & johnson ltd,93,0.685259
24855,Benadryl DR Syrup,dextromethorphan hydrobromide (15mg/5ml),johnson & johnson ltd,123,0.662237
103810,Intadryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,intas pharmaceuticals ltd,38,0.607399
16215,Abidryl-A 1.25mg/15mg/50mg/2.5mg Syrup,terbutaline (1.25mg) ambroxol (15mg),abn drugs,49,0.600239


##### Embedding models naturally give too much importance to brand language but in pharma, Composition = Ground Truth

### Core Idea — Dual-Channel Embeddings

---

### Channel 1 — Brand Semantics

Captures how users speak.

Used for:

- brand names

- natural queries

- conversational search

Example : ***augmentin 625 antibiotic tablet by gsk***

---

#### Channel 2 — Composition Semantics (MOST IMPORTANT)

Captures pharmacological equivalence.

Used for:

- substitute detection

- ingredient similarity

- generic matching

Example: ***active ingredients amoxicillin clavulanate potassium antibiotic combination***

---

In [42]:
def build_brand_doc(row):
    return f"""
    Medicine: {row['cleaned_name']},
    Manufacturer: {row['manufacturer_name']},
    Type: {row['cleaned_name']},
    """

def build_composition_doc(row):
    return f"""
    Active ingredients: {row['composition']}
    Pharmaceutical drug composition.
    """

In [ ]:
dt["brand_text"] = dt.apply(build_brand_doc, axis=1)
dt["composition_text"] = dt.apply(build_composition_doc, axis=1)


NameError: name 'dt' is not defined

In [44]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

brand_embeddings = model.encode(
    dt["brand_text"].tolist(),
    batch_size=64,
    show_progress_bar=True
)

composition_embeddings = model.encode(
    dt["composition_text"].tolist(),
    batch_size=64,
    show_progress_bar=True
)

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 743.56it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
Batches: 100%|██████████| 3807/3807 [09:57<00:00,  6.37it/s]


In [ ]:
import faiss
import numpy as np

dt = pd.read_pickle("../Data/medicine_metadata.pkl")

brand_embeddings = np.array(brand_embeddings).astype("float32")
composition_embeddings = np.array(composition_embeddings).astype("float32")

faiss.normalize_L2(brand_embeddings)
faiss.normalize_L2(composition_embeddings)

TypeError: float() argument must be a string or a real number, not 'IndexFlatIP'

In [49]:
dimension = brand_embeddings.shape[1]

brand_index = faiss.IndexFlatIP(dimension)
brand_index.add(brand_embeddings)

composition_index = faiss.IndexFlatIP(dimension)
composition_index.add(composition_embeddings)

In [50]:
import os

# Define the data directory
DATA_DIR = "../Data"

# 1. Save the Indices
faiss.write_index(brand_index, os.path.join(DATA_DIR, "brand_search.index"))
faiss.write_index(composition_index, os.path.join(DATA_DIR, "composition_search.index"))

# 2. Save the single Metadata file
dt.to_pickle(os.path.join(DATA_DIR, "medicine_metadata_dual.pkl"))

### Can start Fresh Execution from here

In [55]:
import faiss
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")
brand_index = faiss.read_index("../Data/brand_search.index")
composition_index = faiss.read_index("../Data/composition_search.index")
dt = pd.read_pickle("../Data/medicine_metadata_dual.pkl")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 260.31it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [3]:
def format_medicine_results(df_slice):

    results = []

    for _, row in df_slice.iterrows():
        results.append({
            "name": row["name"],
            "composition": row["composition"],
            "manufacturer": row["manufacturer_name"],
            "price": float(row["price"])
        })

    return {
        "type": "medicine_list",
        "results": results
    }

In [4]:
def find_anchor(query):
    q = model.encode([query]).astype("float32")
    faiss.normalize_L2(q)

    scores, indices = brand_index.search(q, 1)
    return dt.iloc[indices[0][0]]

In [5]:
def hybrid_search(query, top_k=5, alpha=0.4):
    """
    alpha controls importance:
    alpha small → composition dominates
    alpha large → brand dominates
    """

    query_embedding = model.encode([query]).astype("float32")
    faiss.normalize_L2(query_embedding)

    brand_scores, brand_ids = brand_index.search(query_embedding, top_k)
    comp_scores, comp_ids = composition_index.search(query_embedding, top_k)

    scores = {}

    # brand contribution
    for score, idx in zip(brand_scores[0], brand_ids[0]):
        scores[idx] = scores.get(idx, 0) + alpha * score

    # composition contribution
    for score, idx in zip(comp_scores[0], comp_ids[0]):
        scores[idx] = scores.get(idx, 0) + (1 - alpha) * score

    ranked = sorted(scores.items(), key=lambda x: x[1], reverse=True)

    final_indices = [idx for idx, _ in ranked[:top_k]]

    df_slice =  dt.iloc[final_indices][
        ["name", "composition", "manufacturer_name", "price"]
    ]

    return format_medicine_results(df_slice)

In [6]:
def find_substitute(query, top_k=5):
    """
    Find substitutes based primarily on composition similarity.

    alpha:
        small  → composition dominates
        large  → brand similarity dominates (optional future use)
    """


    anchor = find_anchor(query)

    if anchor is None:
        return None

    anchor_idx = anchor.name  
    anchor_composition = anchor["composition"]

    comp_text = f"active ingredients {anchor_composition} pharmaceutical drug composition"

    comp_emb = model.encode(
        [comp_text],
        normalize_embeddings=True
    ).astype("float32")

    comp_scores, comp_ids = composition_index.search(comp_emb, top_k + 1)

    scores = {}

    for score, idx in zip(comp_scores[0], comp_ids[0]):
        if idx == anchor_idx:
            continue  # remove same medicine

        scores[idx] = float(score)

    ranked = sorted(scores.items(), key=lambda x: x[1], reverse=True)

    final_indices = [idx for idx, _ in ranked[:top_k]]

    df_slice =  dt.iloc[final_indices][
        ["name", "composition", "manufacturer_name", "price"]
    ]

    return format_medicine_results(df_slice)

In [7]:
find_substitute("benadryl")

{'type': 'medicine_list',
 'results': [{'name': 'New Brethese Syrup',
   'composition': 'diphenhydramine (14.08mg/5ml)  ammonium chloride (138mg/5ml) ',
   'manufacturer': 'galpha laboratories ltd',
   'price': 68.0},
  {'name': 'Intadryl Syrup',
   'composition': 'diphenhydramine (14.08mg/5ml)  ammonium chloride (138mg/5ml) ',
   'manufacturer': 'intas pharmaceuticals ltd',
   'price': 38.0},
  {'name': 'Cofryl Syrup',
   'composition': 'diphenhydramine (14.08mg/5ml)  ammonium chloride (138mg/5ml) ',
   'manufacturer': 'cipla ltd',
   'price': 82.0},
  {'name': 'COF-RYL Cough Syrup',
   'composition': 'diphenhydramine (14.08mg/5ml)  ammonium chloride (138mg/5ml) ',
   'manufacturer': 'cipla ltd',
   'price': 82.0},
  {'name': 'Cofsils Cough Syrup',
   'composition': 'diphenhydramine (14.08mg/5ml)  ammonium chloride (138mg/5ml) ',
   'manufacturer': 'cipla health ltd',
   'price': 95.0}]}

In [8]:
hybrid_search("composition having diphenhydramine", top_k=5)

{'type': 'medicine_list',
 'results': [{'name': 'Bedryl 75mg Soft Gelatin Capsule',
   'composition': 'diphenhydramine (75mg)',
   'manufacturer': 'sun life sciences pvt ltd',
   'price': 50.0},
  {'name': 'Zendryl Syrup',
   'composition': 'diphenhydramine (na)',
   'manufacturer': 'lupin ltd',
   'price': 13.0},
  {'name': 'Exylin 250 DT Tablet',
   'composition': 'diphenhydramine (250mg)',
   'manufacturer': 'win-medicare pvt ltd',
   'price': 43.0},
  {'name': 'Dif 25mg Suspension',
   'composition': 'diphenhydramine (25mg)',
   'manufacturer': 'wallace pharmaceuticals pvt ltd',
   'price': 37.0},
  {'name': 'Benacof 25mg Soft Gelatin Capsule',
   'composition': 'diphenhydramine (25mg)',
   'manufacturer': 'taj pharma india ltd',
   'price': 29.0}]}

Adding Intent Routing

In [17]:
INTENTS = [
    "substitute",
    "general_search"
]

intent_examples = {
    "substitute": [
        "similar to a medicine",
        "alternative to a drug",
        "replacement for a medicine",
        "same composition",
        "cheaper alternative",
        "medicine like this one"
    ],
    "general_search": [
        "medicine information",
        "drug details",
        "medicine manufactured by company",
        "medicine made by company",
        "what is this medicine",
        "find medicines",
        "search medicine",
        "antibiotic tablet",
        "medicine under certain price"
    ]
}

In [18]:
intent_embeddings = {}

for intent, examples in intent_examples.items():
    emb = model.encode(examples)
    intent_embeddings[intent] = emb.mean(axis=0)

In [9]:
def detect_intent(query):

    q = model.encode([query]).astype("float32")
    faiss.normalize_L2(q)

    best_intent = None
    best_score = -1

    for intent, emb in intent_embeddings.items():
        emb = emb.reshape(1, -1).astype("float32")
        faiss.normalize_L2(emb)

        score = (q @ emb.T)[0][0]

        if score > best_score:
            best_score = score
            best_intent = intent

    return best_intent

In [10]:
def handle_query(query):
    
    intent = detect_intent(query)
    print(intent)

    if intent == "substitute":
        print("Called Substitute")
        return find_substitute(query)

    print("Called Hybrid")
    return hybrid_search(query, alpha=0.9)

In [26]:
handle_query("Info about glycomet gp")

general_search
Called Hybrid


{'type': 'medicine_list',
 'results': [{'name': 'Glycomet-GP 0.5 Tablet SR',
   'composition': 'glimepiride (0.5mg)  metformin (500mg)',
   'manufacturer': 'usv ltd',
   'price': 58.0},
  {'name': 'Glycomet-GP 1/850 Tablet SR',
   'composition': 'glimepiride (1mg)  metformin (850mg)',
   'manufacturer': 'usv ltd',
   'price': 75.0},
  {'name': 'Glycomet-GP 2/850 Tablet SR',
   'composition': 'glimepiride (2mg)  metformin (850mg)',
   'manufacturer': 'usv ltd',
   'price': 96.0},
  {'name': 'Glycomet-GP 3/850 Tablet PR',
   'composition': 'glimepiride (3mg)  metformin (850mg)',
   'manufacturer': 'usv ltd',
   'price': 101.0},
  {'name': 'Glycomet 250 Tablet',
   'composition': 'metformin (250mg)',
   'manufacturer': 'usv ltd',
   'price': 15.0}]}

In [27]:
handle_query("cheap alternatives to Glycomet GP")

substitute
Called Substitute


{'type': 'medicine_list',
 'results': [{'name': 'Glimerin-M 1mg/500 mg Tablet',
   'composition': 'glimepiride (1mg)  metformin (500 mg)',
   'manufacturer': 'rhine biogenics pvt ltd',
   'price': 38.0},
  {'name': 'Formistar-PG 1mg/500 mg/15 mg Tablet',
   'composition': 'glimepiride (1mg)  metformin (500 mg) ',
   'manufacturer': 'iva healthcare pvt. ltd.',
   'price': 69.0},
  {'name': 'Formistar G 1mg/500mg Tablet',
   'composition': 'glimepiride (1mg)  metformin (500 mg)',
   'manufacturer': 'iva healthcare pvt. ltd.',
   'price': 40.0},
  {'name': 'Glimizone Plus 1000mg/500mg/15mg Tablet SR',
   'composition': 'glimepiride (1000mg)  metformin (500mg) ',
   'manufacturer': 'carezone healthcare',
   'price': 110.0},
  {'name': 'Dayglim M 0.5mg/500mg Tablet ER',
   'composition': 'glimepiride (0.5mg)  metformin (500mg)',
   'manufacturer': 'daylon healthcare',
   'price': 65.0}]}

In [28]:
def build_prompt(query, data):
    return f"""
    You are a pharmaceutical assistant.

    Use ONLY the provided medicine data.
    Do not add medicines not listed.

    User query:
    {query}

    Medicine data:
    {data}

    Explain clearly:
    - which medicines are equivalent
    - price differences
    - why they are alternatives.
    """

In [11]:
from langchain_core.messages import SystemMessage, HumanMessage

def build_prompt_new(query, data, mode):

    if mode == "substitute":

        system_text = """
You are a pharmaceutical assistant.

Use ONLY the provided medicine data.
Do NOT invent medicines or medical claims.

The medicines listed are substitutes or alternatives
based on similar composition.

Explain clearly:
- which medicines are equivalent substitutes
- important composition similarities
- price differences
- which options may be cheaper
Keep the explanation concise and factual.
"""

    else:  # mode == "info"

        system_text = """
You are a pharmaceutical assistant.

Use ONLY the provided medicine data.
Do NOT invent medicines.

The user wants information about a medicine.

Explain clearly:
- what the medicine is
- its composition
- manufacturer
- price(INR)
If multiple medicines appear, summarize them briefly.
"""

    return [
        SystemMessage(content=system_text),
        HumanMessage(
            content=f"""
User query:
{query}

Medicine database results:
{data}
"""
        )
    ]

In [12]:
from langchain_groq import ChatGroq
from dotenv import load_dotenv
import os

load_dotenv()
os.environ["GROQ_API_KEY"] = os.getenv("GROQ_API_KEY")

In [13]:
llm = ChatGroq(model="llama-3.1-8b-instant")

In [14]:
def call_llm(query, data,intent):
    # prompt = build_prompt(query,data)
    prompt = build_prompt_new(query,data,intent)
    response = llm.invoke(prompt)

    return response

In [15]:
def pharma_assistant(query):

    structured_result = handle_query(query)
    intent = detect_intent(query)
    response = call_llm(query, structured_result,intent)

    return response

In [19]:
pharma_assistant("Info about glycomet gp").pretty_print()

general_search
Called Hybrid
================================== Ai Message ==================================

The information about Glycomet GP is as follows:

- **What the medicine is**: Glycomet GP is a type of oral antidiabetic medication used to treat type 2 diabetes.
- **Its composition**: Glycomet GP contains a combination of two active ingredients:
  - Glimepiride: a sulfonylurea that helps the body produce more insulin.
  - Metformin: a biguanide that helps the body to use insulin more effectively.
- **Manufacturer**: The medicine is manufactured by USV Ltd.
- **Price (INR)**: The prices for different strengths of Glycomet GP are:
  - Glycomet-GP 0.5 Tablet SR: ₹58.0
  - Glycomet-GP 1/850 Tablet SR: ₹75.0
  - Glycomet-GP 2/850 Tablet SR: ₹96.0
  - Glycomet-GP 3/850 Tablet PR: ₹101.0
  - There is also a different medicine called "Glycomet 250 Tablet" which contains only Metformin and is priced at ₹15.0.


In [20]:
pharma_assistant("Alternatives to Glycomet GP").pretty_print()

substitute
Called Substitute
================================== Ai Message ==================================

Based on the medicine database, the alternatives to Glycomet GP are:

- Glimerin-M 1mg/500 mg Tablet
- Formistar-PG 1mg/500 mg/15 mg Tablet
- Formistar G 1mg/500mg Tablet
- Dayglim M 0.5mg/500mg Tablet ER

Important composition similarities among these alternatives are:

- Contain glimepiride and metformin as active ingredients
- Dosage of glimepiride is either 1mg or 0.5mg, which are different but still alternatives
- Dosage of metformin is 500mg in all alternatives, which is the same as Glycomet GP

Price differences are:
- Glimerin-M 1mg/500 mg Tablet: ₹38
- Formistar G 1mg/500mg Tablet: ₹40
- Formistar-PG 1mg/500 mg/15 mg Tablet: ₹69
- Dayglim M 0.5mg/500mg Tablet ER: ₹65
- Glimizone Plus 1000mg/500mg/15mg Tablet SR: ₹110 (not considered as a direct substitute due to different glimepiride dosage)

Options that may be cheaper:
- Glimerin-M 1mg/500 mg Tablet (cheapest option

### Adding context memory

In [21]:
class ConversationMemory:

    def __init__(self):
        self.state = {
            "last_intent": None,
            "last_mode": None,
            "last_results": None
        }

    def update(self, intent=None, mode=None, results=None):

        if intent:
            self.state["last_intent"] = intent

        if mode:
            self.state["last_mode"] = mode

        if results:
            self.state["last_results"] = results

    def get(self):
        return self.state

In [22]:
FOLLOWUP_KEYWORDS = [
    "cheapest",
    "lowest",
    "among them",
    "compare",
    "first",
    "second",
    "which one",
    "under",
    "better"
]



def is_followup(query):
    q = query.lower()
    return any(word in q for word in FOLLOWUP_KEYWORDS)

In [23]:
def handle_followup(query, memory_state):

    results = memory_state["last_results"]

    if not results:
        return None

    q = query.lower()

    if "cheap" in q or "cheapest" in q:

        cheapest = min(results, key=lambda x: x["price"])

        return {
            "followup_type": "followup",
            "action": "cheapest",
            "result": cheapest,
            "results": results
        }
    
    if "compare" in q and len(results) >= 2:

        m1, m2 = results[0], results[1]

        return {
            "followup_type": "comparison",
            "medicine_1": m1,
            "medicine_2": m2,
            "price_difference":
                abs(m1["price"] - m2["price"])
        }

In [31]:
memory = ConversationMemory()

def pharma_assistant(query):

    state = memory.get()

    # ---------- FOLLOWUP ----------
    if is_followup(query) and state["last_results"]:
        structured_result = handle_followup(query, state)
        mode = state["last_mode"]
        print("Structured result is ", structured_result)

    # ---------- NEW QUERY ----------
    else:
        intent = detect_intent(query)

        structured_result = handle_query(query)

        mode = "substitute" if intent == "substitute" else "general"

        memory.update(
            intent=intent,
            mode=mode,
            results=structured_result["results"]
        )

    prompt = build_prompt_new(query, structured_result, mode)

    response = llm.invoke(prompt)

    return response

## Non Agent memory

In [32]:
pharma_assistant("Benadryl syrup").pretty_print()

general_search
Called Hybrid
================================== Ai Message ==================================

Based on the provided information, I have found the results for 'Benadryl syrup' as follows:

There are multiple Benadryl syrups available. Here's a brief summary of each:

1. **Benadryl DR Kids Syrup**: This is a cough syrup containing dextromethorphan hydrobromide (7.5mg/5ml). It is manufactured by Johnson & Johnson Ltd and priced at ₹93.0.
2. **Benadryl DR Syrup**: This is a cough syrup containing dextromethorphan hydrobromide (15mg/5ml). It is manufactured by Johnson & Johnson Ltd and priced at ₹123.0.
3. **Benadryl Syrup**: This is an antihistamine syrup containing diphenhydramine (14.08mg/5ml) and ammonium chloride (138mg/5ml). It is manufactured by Johnson & Johnson Ltd and priced at ₹129.0.

Note that there are two other syrups listed - Abidryl-A 5mg/2mg/10mg/0.5ml Syrup and Abidryl-A 1.25mg/15mg/50mg/2.5mg Syrup, but they do not appear to be related to Benadryl.


In [56]:
pharma_assistant("count of manufactured by Emcure").pretty_print()

general_search
Called Hybrid
================================== Ai Message ==================================

Based on the provided data, there are 4 medicines manufactured by Emcure Pharmaceuticals Ltd. 

Here's the brief information about each of them:

1. **Emduo N Kit**: A combination of lamivudine (150mg) and stavudine (30mg). It's priced at ₹38.0.
2. **Emduo E 30 Kit**: A combination of stavudine (30mg) and lamivudine (150mg). It's priced at ₹75.0.
3. **Emtig 50mg Injection**: Contains tigecycline (50mg). It's priced at ₹5130.0.
4. **Emtri Tablet**: A combination of lamivudine (150mg) and stavudine (40mg). It's priced at ₹1210.0.


In [58]:
pharma_assistant("Glycomet GP information").pretty_print()

general_search
Called Hybrid
================================== Ai Message ==================================

Based on the provided medicine database, here's the information about Glycomet GP:

Glycomet GP is a type of antidiabetic medication. It is available in different strengths and forms.

Here are the details of the available medicines:

- Glycomet GP 0.5 Tablet SR: 
  - Composition: Glimepiride (0.5mg) + Metformin (500mg)
  - Manufacturer: USV Ltd
  - Price: ₹58.0

- Glycomet GP 1/850 Tablet SR: 
  - Composition: Glimepiride (1mg) + Metformin (850mg)
  - Manufacturer: USV Ltd
  - Price: ₹75.0

- Glycomet GP 2/850 Tablet SR: 
  - Composition: Glimepiride (2mg) + Metformin (850mg)
  - Manufacturer: USV Ltd
  - Price: ₹96.0

- Glycomet GP 3/850 Tablet PR: 
  - Composition: Glimepiride (3mg) + Metformin (850mg)
  - Manufacturer: USV Ltd
  - Price: ₹101.0

- Glycomet GP 1 Tablet PR: 
  - Composition: Glimepiride (1mg) + Metformin (500mg)
  - Manufacturer: USV Ltd
  - Price: ₹116.0


In [214]:
pharma_assistant("which is the cheapest?").pretty_print()

Structured result is  {'followup_type': 'followup', 'action': 'cheapest', 'result': {'name': 'Glimerin-M 1mg/500 mg Tablet', 'composition': 'glimepiride (1mg)  metformin (500 mg)', 'manufacturer': 'rhine biogenics pvt ltd', 'price': 38.0}, 'results': [{'name': 'Glimerin-M 1mg/500 mg Tablet', 'composition': 'glimepiride (1mg)  metformin (500 mg)', 'manufacturer': 'rhine biogenics pvt ltd', 'price': 38.0}, {'name': 'Formistar-PG 1mg/500 mg/15 mg Tablet', 'composition': 'glimepiride (1mg)  metformin (500 mg) ', 'manufacturer': 'iva healthcare pvt. ltd.', 'price': 69.0}, {'name': 'Formistar G 1mg/500mg Tablet', 'composition': 'glimepiride (1mg)  metformin (500 mg)', 'manufacturer': 'iva healthcare pvt. ltd.', 'price': 40.0}, {'name': 'Glimizone Plus 1000mg/500mg/15mg Tablet SR', 'composition': 'glimepiride (1000mg)  metformin (500mg) ', 'manufacturer': 'carezone healthcare', 'price': 110.0}, {'name': 'Dayglim M 0.5mg/500mg Tablet ER', 'composition': 'glimepiride (0.5mg)  metformin (500mg)'

In [43]:
def substitute_tool(query: str):
    return find_substitute(query)

def search_tool(query: str):
    return hybrid_search(query,alpha=0.9)

def followup_tool(query: str, memory):
    return handle_followup(query, memory.get())


In [ ]:
AGENT_SYSTEM_PROMPT = """
You are a pharmaceutical assistant planner.

You are NOT allowed to answer user questions.
You must ONLY decide which tool should be used.

Available tools:

1. substitute_tool
   Use when the user asks for:
   - alternatives
   - similar medicines
   - substitutes
   - cheaper versions of a specific medicine

2. search_tool
   Use when the user:
   - asks about a medicine
   - searches for a medicine
   - requests medicine details
   - does general discovery

3. followup_tool
   Use when the user refers to:
   - previous results
   - "among them"
   - "cheapest"
   - "compare these"
   - "which one"
   - any contextual follow-up question

Rules:
- Return ONLY the exact tool name.
- Do NOT explain your choice.
- Do NOT generate any additional text.
"""

In [45]:
def choose_tool(query):

    messages = [
        SystemMessage(content=AGENT_SYSTEM_PROMPT),
        HumanMessage(content=query)
    ]

    decision = llm.invoke(messages).content.strip().lower()

    return decision

In [46]:
def agent_execute(query):

    tool_name = choose_tool(query)
    print("Tool: ",tool_name)
    state = memory.get()

    if tool_name == "followup_tool" and state["last_results"]:
        result = followup_tool(query, memory)
        mode = state["last_mode"]

    elif tool_name == "substitute_tool":
        result = substitute_tool(query)
        mode = "substitute"

    else:
        result = search_tool(query)
        mode = "general"

    # update memory
    memory.update(
        mode=mode,
        results=result.get("results")
    )

    return result, mode

In [47]:
def pharma_assistant(query):

    structured_result, mode = agent_execute(query)

    print(mode)
    
    prompt = build_prompt_new(query, structured_result, mode)

    response = llm.invoke(prompt)

    return response

## Agent Execution

In [57]:
pharma_assistant("Benadryl syrup").pretty_print()

Tool:  search_tool
general
================================== Ai Message ==================================

The user is looking for information about Benadryl syrup. 

Based on the provided medicine database, here's the information about Benadryl syrup:

1. **What the medicine is:** Benadryl syrup is an antihistamine used to relieve symptoms of allergies, cold, and flu.

2. **Composition:**
   - There are two Benadryl syrups with different compositions:
     - **Benadryl DR Kids Syrup:** dextromethorphan hydrobromide (7.5mg/5ml)
     - **Benadryl DR Syrup:** dextromethorphan hydrobromide (15mg/5ml)
     - **Benadryl Syrup:** diphenhydramine (14.08mg/5ml) and ammonium chloride (138mg/5ml)

3. **Manufacturer:** All three Benadryl syrups are manufactured by Johnson & Johnson Ltd.

4. **Price (INR):**
   - **Benadryl DR Kids Syrup:** 93.0
   - **Benadryl DR Syrup:** 123.0
   - **Benadryl Syrup:** 129.0


In [58]:
pharma_assistant("Glycomet gp information").pretty_print()

Tool:  search_tool
general
================================== Ai Message ==================================

The information about Glycomet GP is as follows:

Glycomet GP is a type of medicine that is used to treat Diabetes.

Its composition includes:
- Glimepiride (0.5mg, 1mg, 2mg, or 3mg) - an antidiabetic medication that helps to lower blood sugar levels.
- Metformin (500mg or 850mg) - an antidiabetic medication that helps to control high blood sugar levels.

It is manufactured by USV Ltd.

The prices of Glycomet GP in Indian Rupees (INR) are as follows:
- Glycomet-GP 0.5 Tablet SR: 58.0
- Glycomet-GP 1/850 Tablet SR: 75.0
- Glycomet-GP 2/850 Tablet SR: 96.0
- Glycomet-GP 3/850 Tablet PR: 101.0
- Glycomet-GP 1 Tablet PR: 116.0


In [59]:
pharma_assistant("which is the cheapest?").pretty_print()

Tool:  followup_tool
general
================================== Ai Message ==================================

The cheapest medicine among the options is Glycomet-GP 0.5 Tablet SR. 

Here are the details:

- Medicine: Glycomet-GP 0.5 Tablet SR
- Composition: Glimepiride (0.5mg) + Metformin (500mg)
- Manufacturer: USV Ltd
- Price: 58.0 INR
